File: ruby/atbash-cipher/atbash.rb
module Atbash

  def self.encode(plaintext)
    group_letters(cipher(sanitize(plaintext)))
  end

  private

  PLAIN  = ('a'..'z').to_a.join
  CIPHER = PLAIN.reverse
  IGNORE = /[^a-z\d]/

  def self.sanitize(plaintext)
    plaintext.downcase.gsub(IGNORE, '')
  end

  def self.cipher(sanitized_plaintext)
    sanitized_plaintext.tr(PLAIN, CIPHER)
  end

  def self.group_letters(string)
    string.gsub(/(\w{5})/, '\1 ').strip
  end

end

File: elixir/atbash-cipher/atbash.exs
defmodule Atbash do
  @doc """
  Encode a given plaintext to the corresponding ciphertext

  ## Examples

  iex> Atbash.encode("completely insecure")
  "xlnko vgvob rmhvx fiv"
  """
  @spec encode(String.t) :: String.t
  def encode(plaintext) do
    plaintext
    |> sanitize
    |> cipher
    |> format
  end

  @rejected_characters ~r/[^a-z\d]/

  defp sanitize(string) do
    Regex.replace(@rejected_characters, String.downcase(string), "")
  end

  defp cipher(sanitized_string) do
    to_string(Enum.map to_char_list(sanitized_string), &cipher_char/1)
  end

  @cipher_range ?a..?z

  defp cipher_char(char_code) do
    if Enum.member?(@cipher_range, char_code) do
      @cipher_range.last - (char_code - @cipher_range.first)
    else
      char_code
    end
  end

  defp format(string) do
    String.strip(Regex.replace(~r/(.....)/, string, "\\1 "))
  end

end
